Dockerfile

```diff
@@ -1,111 +1,243 @@
-        libpq5 \
-        imagemagick \
-        ffmpeg \
-        libjemalloc2 \
-        libicu72 \
-        libidn12 \
-        libyaml-0-2 \
-        file \
-        ca-certificates \
-        tzdata \
-        libreadline8 \
-        tini && \
-    ln -s /opt/mastodon /mastodon && \
-    corepack enable
-
-# Note: no, cleaning here since Debian does this automatically
-# See the file /etc/apt/apt.conf.d/docker-clean within the Docker image's filesystem
-
-COPY --chown=mastodon:mastodon . /opt/mastodon
-COPY --chown=mastodon:mastodon --from=build /opt/mastodon /opt/mastodon
-
-ENV RAILS_ENV="production" \
-    NODE_ENV="production" \
-    RAILS_SERVE_STATIC_FILES="true" \
-    BIND="0.0.0.0" \
-    MASTODON_VERSION_PRERELEASE="${MASTODON_VERSION_PRERELEASE}" \
-    MASTODON_VERSION_METADATA="${MASTODON_VERSION_METADATA}"
-
-# Set the run user
-USER mastodon
-WORKDIR /opt/mastodon
+# Copy bundler and node packages from build layer to container
+COPY --from=build-bundler /opt/mastodon /opt/mastodon/
+COPY --from=build-bundler /usr/local/bundle/ /usr/local/bundle/
+COPY --from=build-node /opt/mastodon /opt/mastodon/
+
+ARG TARGETPLATFORM
+
+RUN \
+# Use Ruby on Rails to create Mastodon assets
+  OTP_SECRET=precompile_placeholder SECRET_KEY_BASE=precompile_placeholder bundle exec rails assets:precompile; \
+# Cleanup temporary files
+  rm -fr /opt/mastodon/tmp;
 
-# Precompile assets
-RUN OTP_SECRET=precompile_placeholder SECRET_KEY_BASE=precompile_placeholder rails assets:precompile
+# Prep final Mastodon Ruby layer
+FROM ruby
+
+ARG TARGETPLATFORM
+
+# hadolint ignore=DL3008
+RUN \
+# Mount Apt cache and lib directories from Docker buildx caches
+--mount=type=cache,id="apt-cache-${TARGETPLATFORM}",target=/var/cache/apt,sharing=locked \
+--mount=type=cache,id="apt-lib-${TARGETPLATFORM}",target=/var/lib/apt,sharing=locked \
+# Mount Corepack and Yarn caches from Docker buildx caches
+--mount=type=cache,id="corepack-cache-${TARGETPLATFORM}",target=/usr/local/share/.cache/corepack,sharing=locked \
+--mount=type=cache,id="yarn-cache-${TARGETPLATFORM}",target=/usr/local/share/.cache/yarn,sharing=locked \
+# Apt update install non-dev versions of necessary components
+  apt-get install -y --no-install-recommends \
+    libssl3 \
+    libpq5 \
+    libicu72 \
+    libidn12 \
+    libreadline8 \
+    libyaml-0-2 \
+  ;
+
+# Copy compiled assets to layer
+COPY --from=build-assets /opt/mastodon/public/packs /opt/mastodon/public/packs
+COPY --from=build-assets /opt/mastodon/public/assets /opt/mastodon/public/assets
+# Copy bundler components to layer
+COPY --from=build-bundler /opt/mastodon/ /opt/mastodon/
+COPY --from=build-bundler /usr/local/bundle/ /usr/local/bundle/
+
+RUN \
+# Test ImageMagick and ffmpeg availablity
+  convert -version; \
+  ffmpeg -version; \
+  ffprobe -version; \
+# Pre-create and chown system volume to Mastodon user
+  mkdir -p /opt/mastodon/public/system; \
+  chown mastodon:mastodon /opt/mastodon/public/system;
```

Comment:
Should we pre-compile bootsnap cache here as well, like in the Rails upstream Dockerfile?

Reply:
Ah that must not have been committed while I was testing. I've had to replacement my dev box a couple times since we looked at that. Will fix.